bpf/bpf_host.c

```diff
@@ -1308,12 +1308,46 @@ int cil_from_host(struct __ctx_buff *ctx)
 	return handle_netdev(ctx, true);
 }
 
+#if defined(ENABLE_IPSEC) && defined(ENABLE_IPV4) && defined(TUNNEL_MODE)
+/*
+ * If the traffic is indeed overlay traffic and it should be encrypted
+ * CTX_ACT_REDIRECT is returned, unless an error occurred, and the caller can
+ * return this code to TC.
+ *
+ * CTX_ACT_OK is returned if the traffic should continue normal processing.
+ *
+ * IS_ERR can be used to determine if this function ran into an error.
+ */
+static __always_inline int do_encrypt_overlay(struct __ctx_buff *ctx,
+					      __u16 proto)
+{
+	int ret = CTX_ACT_OK;
+	struct iphdr __maybe_unused *ipv4;
+	void __maybe_unused *data, *data_end = NULL;
+
+	/* only support vxlan for now */
+	if (TUNNEL_PROTOCOL != TUNNEL_PROTOCOL_VXLAN)
+		return ret;
```

Comment:
`build_bug_on()` ?

Reply:
Could this get annoying when TUNNEL_MODE is on but TUNNEL_PROTOCOL is a protocol other then VXLAN? I'm worried that building tests or building in order to check for compilation issues may break.

Reply:
Not if all this code is behind a flag (e.g., `ENABLE_ENCRYPTED_OVERLAY`).

Reply:
I'm not following. How will we control if the overlay is encrypted or not? We don't want to enable this for existing users.

Reply:
Encrypted Overlay is 'opt-in' by the datapath right? Specific code needs to place the 'ENCRYPTED_OVERLAY_ID' in the tunnel_key.id to inform the DP that the traffic should be "encrypted overlay" traffic. Without this, nothing changes for any traffic in the cluster.

Reply:
OK, I think I understand now. You're saying that a consumer of encrypted overlay like multicast will know to set the new security ID. That's good, but wasn't the goal here to also enable the usual IPsec from using encrypted overlay? If we want to do that, I don't really see a way besides having a new flags (in agent and datapath).

Reply:
I think Paul is spot-on. There's no good reason why this code-path should be enabled by default - it adds BPF complexity, and exposes all existing users to potential bugs (that would otherwise only hit the multicast use-case).

Reply:
> That's good, but wasn't the goal here to also enable the usual IPsec from using encrypted overlay?

I'm not sure that's the case. I think the use-case here is to allow traffic which cannot enter XFRM hooks normally to have an alternative method to do so. I intended for this feature to run along-side normal IPSec encryption flow.

Reply:
Discussion with @pchaigno @julianwiedmann made it clear why a flag would be appropriate for this feature. 
I'll add this in with the next commit.

Reply:
Added a new 'encryption.ipsec.encryptedOverlay` (agent) and ENABLE_ENCRYPTED_OVERLAY (datapath) with latest push. Resolving.